packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
+/// the current set of Widget states.
+///
+/// {@template flutter.widgets.WidgetStateProperty.WidgetStateMap}
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.map(<WidgetStateMapKey, Color?>{
+///   WidgetState.error: Colors.red,
+///   WidgetState.hovered & WidgetState.focused: Colors.blueAccent,
+///   WidgetState.focused: Colors.blue,
+///   ~WidgetState.disabled: Colors.black,
+/// });
+///
+/// // the same implementation, but with a WidgetPropertyResolver:
+/// WidgetStateProperty.resolveWith<Color?>((Set<WidgetState> states) {
+///   if (states.contains(WidgetState.error)) {
+///     return Colors.red;
+///   } else if (states.contains(WidgetState.hovered) && states.contains(WidgetState.focused)) {
+///     return Colors.blueAccent;
+///   } else if (states.contains(WidgetState.focused)) {
+///     return Colors.blue;
+///   } else if (!states.contains(WidgetState.disabled)) {
+///     return Colors.black;
+///   }
+///   return null;
+/// });
+/// ```
+///
+/// A widget state combination can be stored in a variable,
+/// and [WidgetState.any] can be used for non-nullable types:
+///
+/// ```dart
+/// final WidgetStateMapKey selectedError = WidgetState.selected & WidgetState.error;
+///
+/// final WidgetStateProperty<Color> color = WidgetStateProperty<Color>.map(
+///   <WidgetStateMapKey, Color>{
+///     selectedError & WidgetState.hovered: Colors.redAccent,
+///     selectedError: Colors.red,
+///     WidgetState.any: Colors.black,
+///   },
+/// );
+///
+/// // WidgetPropertyResolver implementation:
+/// final WidgetStateProperty<Color> colorResolveWith = WidgetStateProperty.resolveWith<Color>(
+///   (Set<WidgetState> states) {
+///     if (states.containsAll(<WidgetState>{WidgetState.selected, WidgetState.error})) {
+///       if (states.contains(WidgetState.hovered)) {
+///         return Colors.redAccent;
+///       }
+///       return Colors.red;
+///     }
+///     return Colors.black;
+///   },
+/// );
+/// ```
+/// {@endtemplate}
+typedef WidgetStateMap<T> = Map<WidgetStateMapKey, T>;
+
+// A private class, used to create the [WidgetStateProperty.map] constructor.
+class _WidgetStateMapper<T> implements WidgetStateProperty<T> {
+  const _WidgetStateMapper(this.map);
+
+  final WidgetStateMap<T> map;
+
+  @override
+  T resolve(Set<WidgetState> states) {
+    for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
+      if (entry.key.matchesSet(states)) {
+        return entry.value;
+      }
+    }
+
+    late final String errorMessage = 'The current set of material states is $states.\n'
+        'None of the provided map keys matched this set, and the type "$T" is non-nullable.';
+    assert(
+      null is T,
+      '$errorMessage\n'
+      'Consider using "WidgetStateProperty<$T?>.map()", '
+      'or adding the "WidgetState.any" key to this map.',
+    );
+
+    try {
+      return null as T;
+    } on TypeError {
+      throw ArgumentError(errorMessage);
+    }
+  }
```

Comment:
I wonder if we shouldn't allow implicitly returning null. Would it maybe be better to explicitly require an entry for null in the map if you want it to ever resolve to null?

Would be interesting to look at some use cases to see how common the null case is and if making it explicit would just add a ton of boilerplate.

Reply:
Also, this implicit null behavior needs to be documented somewhere clearly.

Reply:
> this implicit null behavior needs to be documented somewhere clearly.

Absolutely!